### DISCUSSION TITLE
GraphMask Implementation for the main PyG library

### DISCUSSION THREAD

**Discussion by fork123aniket** [pyg-team/pytorch_geometric#7271] [Ideas]
GraphMask Explainability technique is appearing in the `torch_geometric.contrib.explain` package. @rusty1s Any idea as to how to get this implementation to completion so that we can have it in the `main` PyG library??

To make the context more clear, the current version of GraphMask runs successfully for three different `GNN` layers: `GCNConv`, `GATConv`, and `FastRGCNConv` on three different `GNN` tasks: `Node Classification`, `Graph Classification`, and `Link Prediction`.

Thank you.

**Comment by rusty1s** [comment]
Sorry for my late reply here. Is there a problem with this being in `contrib`? I don't see this package as less valuable, we just prefer to keep complex external implementations there. If this is a problem, let me know and I will take a look to see what needs to be done to move this to `torch_geometric.explain`.

**Reply by fork123aniket** [reply]
@rusty1s Yes. Just want this implementation to be included in the main library (`torch_geometric.explain`). Besides, have made two changes to the existing `GraphMask` Implementation present in `torch_geometric.contrib`: (i) made the implementation `layer_type` independent so that the use of the `GraphMask Explainer` does not require the user to type in the value of `layer_type` parameter for the original `GNN` model being explained; (ii) the method of setting the `node` as well as `edge` masks does not have to rely on the `layer_type` of the `GNN` model as dimensions of masks are inferred automatically. Moreover, tested the `GraphMask` implementation with these two changes for two different GNN layers ` SimpleConv` and `SAGEConv` that resulted in the successful generation of `node` and `edge` masks.

Please let me know your thoughts on this and if raising a new pull request with the changes made would help…

**Reply by rusty1s** [reply]
Super, do you want to send two PRs for this? Dropping the `layer_type` would be indeed a requirement to move this to `torch_geometric.explain`.

**Reply by fork123aniket** [reply]
@rusty1s Please take a look at PR #7445. The code for this PR reflects the changes made to make the implementation `layer_type` independent. Please do let me know if anything else is required to ease the review process of this PR. Thanks!!!

**Reply by rusty1s** [reply]
Super, will take a look.

**Reply by fork123aniket** [reply]
@rusty1s Thanks so much for merging the pull request yesterday. Really appreciate this move!! However, I do see the implementation is still in `torch_geometric.contrib.explain`. May I ask what are the expectations that the existing `GraphMask Explainer` implementation requires to meet in order to have it included in the main `torch_geometric.explain` package??

**Reply by rusty1s** [reply]
Hey @fork123aniket, yes I needed to move it back in order to see the changes/diff made in this PR. Otherwise, this was not possible. Next step would probably to add some more documentation to functions and add type hints to their arguments.

**Reply by fork123aniket** [reply]
Hi @rusty1s . As advised above, have included details in more functions and added `datatype hints` to the functions' arguments. Please consider taking a look at the PR #7503 and see if this `GraphMaskExplainer` implementation now can be included in the main library (`torch_geometric.explain`) or do let me know if something is still missing which needs to be included before driving this implementation from `torch_geometric.contrib.explain` to `torch_geometric.explain`. Thanks!!

**Reply by fork123aniket** [reply]
Hi @rusty1s. Now that we see PR #7504 has been merged, may I know what all expectations are left that need to be fulfilled that could aid in driving the implementation from `torch_geometric.contrib.explain` to `torch_geometric.explain`?? Could you shed some light on this, please??

**Comment by fork123aniket** [comment]
Hi @rusty1s, any update on this?? Can you guide me as to what all changes are yet to be made to make the transition of `GraphMask Explainer` from `torch_geometric.contrib.explain` to `torch_geometric.explain`?? Your advice concerning this transition may indeed help expedite this integration of `GraphMask Explainer` into the main library.

Thank you. Looking forward to hearing from you very soon.

**Reply by rusty1s** [reply]
I think this indeed looks much better now. There are still a few things I would like to iron out, e.g.:
* `explain_message` creates new modules (e.g., `LayerNorm`) inside each function call. I don't think this is necessary.
* There is no reason the `GraphMaskExplainer` re-implements the loss functions such as `_loss_binary_classification`, etc methods from the base class

**Reply by fork123aniket** [reply]
1. Currently, each `explain_message()` function call generates `layer-wise stochastic gates` (`random edge weights`) at the expense of the creation of two modules (`LayerNorm` and `ReLU`). After each message passing layer, the `explainer` must generate the `edge weights` using these two modules at least as per the approach described in the paper. It is possible that I ain't familiar with a probably easier  workaround for this, hence feel free to recommend a simple, yet efficient approach to use these two modules while making sure of not having these employed at every single function call. Besides, one possible solution I can think of is to normalize the `layer weights` inside `torch_geometric.nn.conv.MessagePassing` itself just before enabling `explain_message()` (present inside `torch_geometric.contrib.explain.GraphMaskExplainer`) for each message passing layer, however, ain't aware of how efficient this solution could be. Nevertheless, am open to removing these two modules altogether as long as this allows `GraphMaskExplainer` to work properly. What's your advice on this??

2. Affirmative!! I accidentally overlooked the `base explainer algorithm` implementation while implementing `GraphMaskExplainer` and eventually found three functions (`_loss_binary_classification`, `_loss_multiclass_classification`, and `_loss_regression`) being duplicated in the current `GraphMaskExplainer` implementation. Based on your advice, I have removed all these three and will soon raise a PR for merging purposes.

**Reply by rusty1s** [reply]
1. Yes, I understand that. I am just arguing that we don't need to create this modules. We can use `x.relu()` or `x.clamp(min=0)` to replace the `ReLU` module, and we can use `torch.linalg.norm` to replace the `LayerNorm` (its weights are set to `1` and `0` anyway).
2. Super, sounds good :)

**Reply by fork123aniket** [reply]
The purpose of using the `LayerNorm` module is to apply per-element normalization. Despite various attempts, couldn't find `torch.linalg.norm`'s arguments' values in such a way that it provides output with shape `(N, *)` considering the same input shape. Rather, I did the following to have the output tensor with the same shape as that of the input's, yet achieved per-element normalization:-
```
norm = Sequential(LayerNorm(out.size(-1)).to(out.device), ReLU())
basis_messages = norm(out)
```
altered to
```
basis_messages = torch.nn.functional.layer_norm(out, (out.size(0), out.size(-1))).to(out.device).relu()
```
So, can we use `torch.nn.functional.layer_norm` instead of `torch.linalg.norm`, given both input (`out`) and output (`basis_messages`) tensors must have the same shape??

**Reply by rusty1s** [reply]
Yes, let's use `torch.nn.functional.layer_norm` instead, this would be great :)

**Reply by fork123aniket** [reply]
Done!! If no more changes are required, shall I tweak the code to ease the integration of the current `GraphMask` implementation into `torch_geometric.explain` from `torch_geomtric.contrib.explain` just by changing the paths in the code and attach other relevant code files (`example.py` and `test.py`) to the upcoming PR for final review?? What do you recommend??

**Reply by rusty1s** [reply]
Thanks. Can you send a PR with the corresponding changes? Let's do the final migration in a follow-up.

**Reply by fork123aniket** [reply]
As advised, have just opened a follow-up PR (#7779) for this final migration of the `GraphMaskExplainer` implementation. Please consider giving it a review.
